ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs

```diff
@@ -0,0 +1,124 @@
+import           Test.QuickCheck.Extras (unsafeMapSuchThatJust)
+import           Test.QuickCheck.Random (QCGen)
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.QuickCheck (le)
+import           Test.Util.TestBlock (TestHash (TestHash))
+import           Test.Util.TestEnv (adjustQuickCheckTests)
+
+tests :: TestTree
+tests =
+  adjustQuickCheckTests (`div` 10) $
+  testProperty "uniform random schedule" prop_uniformSchedule
+
+genChainsAndSchedule :: PointScheduleConfig -> ScheduleType -> QC.Gen (GenesisTest, PointSchedule)
+genChainsAndSchedule scheduleConfig scheduleType =
+  unsafeMapSuchThatJust do
+    numAdversaries <- QC.choose (1, 4)
+    gt <- genChains numAdversaries
+    seed :: QCGen <- arbitrary
+    pure ((gt,) <$> runSTGen_ seed (\ g -> genSchedule g scheduleConfig scheduleType (gtBlockTree gt)))
+
+simpleHash ::
+  HeaderHash block ~ TestHash =>
+  ChainHash block ->
+  [Word64]
+simpleHash = \case
+  BlockHash (TestHash h) -> reverse (NonEmpty.toList h)
+  -- not matching on @GenesisHash@ because 8.10 can't prove exhaustiveness of
+  -- TestHash with the equality constraint
+  _ -> []
+
+makeProperty ::
+  GenesisTest ->
+  Int ->
+  StateView ->
+  [PeerId] ->
+  Property
+makeProperty genesisTest advCount StateView {svSelectedChain} killed =
+  classify genesisWindowAfterIntersection "Full genesis window after intersection" $
+  classify (isOrigin immutableTipHash) "Immutable tip is Origin" $
+  classify (length killed > 0) "Some peers were killed" $
+  classify (length killed == advCount) "All adversaries were killed" $
+  classify (advCount < length (btBranches (gtBlockTree genesisTest))) "Some adversaries performed rollbacks" $
+  counterexample killedPeers $
+  conjoin [
+    property (isHonest immutableTipHash),
+    immutableTipIsRecent
+  ]
+  where
+    immutableTipIsRecent =
+      counterexample ("Age of the immutable tip: " ++ show immutableTipAge) $
+      immutableTipAge `le` s + fromIntegral d + 1
+
+    SlotNo immutableTipAge = case (honestTipSlot, immutableTipSlot) of
+      (At h, At i)   -> h - i
+      (At h, Origin) -> h
+      _              -> 0
+
+    isOrigin = null
+
+    isHonest = all (0 ==)
+
+    immutableTipHash = simpleHash (AF.anchorToHash (AF.anchor svSelectedChain))
+
+    immutableTipSlot = AF.anchorToSlotNo (AF.anchor svSelectedChain)
+
+    killedPeers = case killed of
+      [] -> "No peers were killed"
+      peers -> "Some peers were killed: " ++ intercalate ", " (condense <$> peers)
+
+    GenesisWindow s = gtGenesisWindow genesisTest
+
+    Delta d = gtDelay genesisTest
+
+    honestTipSlot = AF.headSlot (btTrunk (gtBlockTree genesisTest))
+
+    Classifiers {genesisWindowAfterIntersection} = classifiers genesisTest
+
+prop_uniformSchedule :: QC.Gen QC.Property
+prop_uniformSchedule = do
+  (genesisTest, schedule) <- genChainsAndSchedule scheduleConfig Uniform
+  pure $ withMaxSuccess 10 $
```

Comment:
```suggestion
  pure $
```
This interferes with the call to `adjustQuickCheckTests` otherwise.

Reply:
It does indeed, as in even in CI and nightly, we will run only 10 tests. However, what we noticed since is that `adjustQuickCheckTests` and `withMaxSuccess` do not behave the same with respect to the size of inputs. It is just a guess but I think that QuickCheck follows a curve somewhere to increase the number of ran test cases. With `withMaxSuccess`, this curve is just cut after 10 successes, while with `adjustQuickCheckTests` it is scaled to the new number. This means that a test with ``adjustQuickCheckTests (`div` 10)`` expecting 10 successes will still run with much bigger sizes than one with `withMaxSuccess 10`, and therefore will take more time. Not sure what's the best way to approach this, though.

Reply:
Yes, in any case, `withMaxSuccess` is bad and we should avoid it, because it doesn't allow any override or anything.